### ISSUE TITLE
Loops in slang (DSL)

### ISSUE THREAD

**Issue by orius123** [openscore/score-language#30]
Ok, so I want ti start with defining how loops will look in our slang.

Lets take this basic flow as an example

```yaml
flow:
  inputs: 
    - list_input: [1,2,3]
    - normal_input
  workflow:
    iter_task:
      do:
        ops.print:
          - list_input
          - normal_input
```

I was thinking of few ways we can go:
Currently those two ways don't give a way to aggregate the result,
and therefore they are not so useful because you can't publish to next tasks. 
(you can still make void action for each though)

**For..do**
```yaml
iter_task:
  # We can use another word of course like foreach..
  for: input in list_input
  do: 
    ops.print:
      - input
      - normal_input
```
Pros:
- verbose
- clean

Cons:
- verbose

**For inline**
```yaml
iter_task:
  do: 
    ops.print:
      - input in list_input
      - normal_input
```
Pros:
- less to write
Cons:
- theoretically i can iterate over more than one input (matrix)

Lets here you.
@meshipeer @oritstone @meirwah @avimora @TudorLesanHP @Bonczidai @lrevnic @avigail-oron

which one did you like?
do you have another suggestion?
do you think we need to give an option to stop the iteration according to a condition? (while style..)
do you think we need to aggregate the result? and how?

**Comment by oritstone** [comment]
- I prefer the first option - it is much clearer that we are looping. Maybe we need to indent the do: inside
- I think we need to add also conditional loops
- I think we should add a way to aggregate the result. What about doing the for loop on the entire task, and this way use the publish?

**Comment by meirwah** [comment]
I prefer the first option too

**Comment by meshipeer** [comment]
i also prefer the first, but i have few questions:
1. did we take a look how other DSLs define loops?
2. can we break the loop in the middle?
3. how we solve the use case we want to wait for something to finish (while...)?

**Comment by avigail-oron** [comment]
I prefer the first one too.
we must provide a way to aggregate the result. breaking out is very nice to have IMO.

**Comment by orius123** [comment]
considering your inputs, lets try again:
I don't like it very much, but this is the best I got so far..
what say you?

**Given the flow:**
```yaml
flow:
  inputs: 
    - list_input: [1,2,3]
    - normal_input
  workflow:
    iter_task:
      do:
        ops.print:
          - list_input
          - normal_input
```
**And operation**
```yaml
- some_op
    inputs:
      - iterating_input
      - normal_input
    action: 
      ...
    outputs:
      - op_result
      - status_code
    results
      - SUCCESS: int(status_code) == 1
      - OTHER
```


**For..do**
```yaml
iter_task:
  # We can use another word of course like foreach..
  for: input in list_input
  do: 
    ops.print:
      - iterating_input: input
      - normal_input
  #we run publish after every iteration
  #aggregate
  publish:
    - agg_result: agg_result + op_result
    - agg_status: agg_status && (status_code == 1)
  
  #we run navigate after every iteration as well
  # default success navigates back to the step with the next index
  # other navigations can break
  navigate:
    OTHER: step_56

**Comment by okitac** [comment]
So --  I think your flow comes up with (it's doubtless clear to you what it does, but I'm still geting used to slang):
* list_input[1]
* normal_input
* list_input[2]
* normal_input
* list_input[3]
* normal_input

Given that, it would be much clearer to use
```yaml
for:  input in list_input
```
rather than the implicit list input (that, or I'm missing something that's obvious to you about what you're doing there).

I'd hope that you don't have to aggregate results (that you can either pull the results from that iteration, or the results when you exit the loop) -- that would make it challenging to nest loops, which is something that's exceedingly likely.

```yaml
navigate:
    OTHER:  step_56
```
looks an awful lot like 'GOTO' -- having an explicit break and continue would be clearer and preferable IMO, although it's certainly possible to get the same effect with different navigate options.

**Comment by avigail-oron** [comment]
Note that the “goto” is anyway required to navigate the tree, (in case you do not wish to proceed to the next step) so it does make sense to allow it here.
I agree, that for the default behavior it makes sense to have a “break” which is actually a “go to the next step as define in the file”

**Comment by orius123** [comment]
Ok, so lest start from the end..
this how we propose that  a simple loop over a list will look like.

**Simple example**
```yaml
workflow:
  print_hosts:
    for: host in hosts
      do:
        ops.print:
          - host
          - port: default_port
  task2:
    .........
```
----------------------------------
And now lets dig into it.
Our example is based on the given flow and operation.

**Operation**
```yaml
operation:
  name: print
  inputs:
    - host
    - port
  action:
    python_script: |
      ...
  outputs:
    - op_result
    - status_code
  results:
    GOOD: host == 2
    BAD: host == 3
    SUCCESS
```

**Flow**
```yaml
flow:
  name: some_flow
  inputs:
    - hosts: [1,2,3]
    - default_port: "'8080'"
  workflow:
    ...
```

We suggest to add the `for` keyword to our language
(later we plan to add `while` and more variations of `for` loops as well)

This is the complete example, please take a look and try to understand it **before** reading the detailed explanation:
```yaml
print_hosts:
  for: host in hosts
    #delimiter: ","
    do:
      ops.print:
        - host
        - port: default_port
    publish:
      - agg_result: agg_result + op_result
      - agg_status: agg_status && (status_code == 1)
    break:
      - BAD
  navigate:
    GOOD: task3
    BAD: task2
task2:
  .........
task3:
  .........
task4:
```
---------------------
## Let's explain one by one.
### `for: host in hosts`
* First, we have a new keyword in the language `for`.
* Second, we have a new syntax `x in [x1,x2,..xn]`
  this syntax is similar to `Python` for loop over a collection, but it's pretty common in other    languages as well.

  ```ptyhon
      collection = ['hey', 5, 'd']
      for x in collection:
        print x
  ```

  > we do think of supporting more variants, like `x in range()` ... ,and `while` loops, but this not first    priority IMHO.


### **Optional** `delimiter` propety
  If the user passes a `String` with delimiter separated inputs instead of a list (we think it will be useful because the CLI works only with Strings currently), then the user can specify a delimiter to split the string by. 

 By default if the user passes a `String` w/o specifying a `delimiter` we will use a `comma` "," to split.

### Regular `task` keywords indentation
  We indent the `do` and `publish` - `task` and the new keyword `break` under the `for` keyword, to imply the user that they will happen for each iteration of the loop.

### `do` 
  I think this one is pretty straight forward, you just write the same `do` you are used to, from a regular task, but now, it will run on each iteration.

### `publish`
  Here it's a bit more tricky, the `publish` will run after each iteration, and will update the flow context with the results. This way we allow the user to aggregate the date from the loop, so he can handle it in a different task.

### `break`
  This is a property, that enables the user to break the loop in given results. after breaking the execution will go tot the `navigate` definition of the task. 

**Default** - The default behavior of `break` is that it breaks for a `FAILURE` result, that way we try to be consistence with the `FAILURE` result in a regular task.
The user can override it by giving an empty list -  `break: []`.

### `navigate`
  You can see that the navigate indentation is on the task properties (same level as the `for` keyword), the reason for that is because the `navigate` happens only after the loop has ended, or broke. the value to navigate by will be the value of the last iteration.

**Default** - The default behavior of the `navigate` is like in a regular task.
* `SUCCESS` - goes to next step
* `FAILURE` - fails the flow, or go to the `on-failure` if exists.

@openscore/slang-functional-reviewers

WDYT?

 @michalhugim @ziso @tomershadi @avigail-oron, please accept the invitation so we won't have to manually add you each time..

**Comment by orius123** [comment]
Hi, 

went on and implemented. you can see implementation under the loops branch, and will be merged later on.

only change is that we support expression as the coolection in the `for`, so the user can call functions
for example:
`for: x in range(0, 9)'
or
`for: x in xrange(0, 9)'
or 
`for: host in "host1;host2".split(";")'

therefore we didn't added the `delimiter` property. 
Still strings are being split by comma.

closing..
3liya
